File: tools/verify_offline.py
#!/usr/bin/env python3
"""
==============================================================================
ScatterID — "Don't Trust, Verify" Standalone Offline CLI Verifier (Python)
==============================================================================
This tool allows third-party auditors and verifiers to mathematically verify
the cryptographic authenticity of an issued ScatterID credential COMPLETELY
OFFLINE:
  1. Zero-Knowledge Pre-image Commitment (RFC 8785 + FIPS 202 SHA3-256)
  2. NIST FIPS 204 ML-DSA-65 Post-Quantum Digital Signature Verification

Ecosystem Role:
  - Python CLI tool for security auditors, relying parties, and data science pipelines.
  - For Node.js / TypeScript environments, use: node tools/verify_offline.js

Usage:
  python3 tools/verify_offline.py <path-to-credential.json> [--public-key <hex>]
  cat credential.json | python3 tools/verify_offline.py
==============================================================================
"""

import sys
import os
import json
import hashlib
import hmac
import math
import re

def _format_float_ecma(f: float) -> str:
    if math.isnan(f) or math.isinf(f):
        raise ValueError("NaN and Infinity are not allowed in RFC 8785 JSON")
    if f == 0:
        return '0'
    if f < 0:
        return '-' + _format_float_ecma(-f)

    stringified = str(f)
    exponent_str = ''
    exponent_value = 0
    q = stringified.find('e')
    if q > 0:
        exponent_str = stringified[q:]
        if exponent_str[2:3] == '0':
            exponent_str = exponent_str[:2] + exponent_str[3:]
        stringified = stringified[0:q]
        exponent_value = int(exponent_str[1:])

    first = stringified
    dot = ''
    last = ''
    q = stringified.find('.')
    if q > 0:
        dot = '.'
        first = stringified[:q]
        last = stringified[q + 1:]

    if last == '0':
        dot = ''
        last = ''

    if 0 < exponent_value < 21:
        first += last
        last = ''
        dot = ''
        exponent_str = ''
        q = exponent_value - len(first)
        while q >= 0:
            q -= 1
            first += '0'
    elif -7 < exponent_value < 0:
        last = first + last
        first = '0'
        dot = '.'
        exponent_str = ''
        q = exponent_value
        while q < -1:
            q += 1
            last = '0' + last

    return f'{first}{dot}{last}{exponent_str}'


def _fallback_jcs(obj):
    """Zero-dependency RFC 8785 JSON Canonicalization Scheme (JCS) serializer."""
    if obj is None:
        return 'null'
    if isinstance(obj, bool):
        return 'true' if obj else 'false'
    if isinstance(obj, int):
        return str(obj)
    if isinstance(obj, float):
        return _format_float_ecma(obj)
    if isinstance(obj, str):
        return json.dumps(obj, ensure_ascii=False)
    if isinstance(obj, (list, tuple)):
        return '[' + ','.join(_fallback_jcs(x) for x in obj) + ']'
    if isinstance(obj, dict):
        # RFC 8785 §3.2.3: object keys sorted by UTF-16 code unit representation
        sorted_keys = sorted(obj.keys(), key=lambda k: k.encode('utf-16be'))
        return '{' + ','.join(json.dumps(k, ensure_ascii=False) + ':' + _fallback_jcs(obj[k]) for k in sorted_keys) + '}'
    raise TypeError(f"Type {type(obj)} is not serializable under RFC 8785")

try:
    import rfc8785
    def canonicalize(obj):
        """RFC 8785 deterministic JSON representation with IEEE 754 number serialization."""
        return rfc8785.dumps(obj).decode('utf-8')
except ImportError:
    def canonicalize(obj):
        """Zero-dependency RFC 8785 fallback with UTF-16 code unit key sorting."""
        return _fallback_jcs(obj)

import argparse

# ANSI terminal colors
BOLD = '\033[1m'
GREEN = '\033[32m'
RED = '\033[31m'
CYAN = '\033[36m'
YELLOW = '\033[33m'
MAGENTA = '\033[35m'
RESET = '\033[0m'

def print_header():
    print(f"{BOLD}{CYAN}======================================================================{RESET}")
    print(f"{BOLD}{CYAN}      ScatterID — Independent Offline Cryptographic Verifier (Python) {RESET}")
    print(f"{BOLD}{CYAN}======================================================================{RESET}")
    print("Standards: RFC 8785 JSON Canonicalization (JCS) | SHA3-256 | ML-DSA-65 (FIPS 204)")
    print(f"Mode:      {BOLD}100% OFFLINE (Zero Network Transit){RESET}\n")


def verify_offline(raw_json, cli_public_key=None):
    try:
        data = json.loads(raw_json)
    except Exception as e:
        print(f"{RED}[ERROR] Invalid JSON: {e}{RESET}", file=sys.stderr)
        sys.exit(1)

    cred = data.get('credential', data)

    raw_claim = cred.get('rawClaim') or cred.get('claim')
    salt_hex = cred.get('salt')
    expected_hash = cred.get('dataHash')
    signature_hex = cred.get('signature')
    public_key_hex = cli_public_key or cred.get('publicKey') or cred.get('publicKeyHex')
    public_key_id = cred.get('publicKeyId', 'N/A')
    credential_id = cred.get('credentialId') or cred.get('id') or 'N/A'
    algorithm = cred.get('algorithm', 'ML-DSA-65')
    anchor_tx = cred.get('anchorTxId', 'None')

    if not raw_claim or not salt_hex or not expected_hash:
        print(f"{RED}[ERROR] Incomplete credential object. Required: 'rawClaim', 'salt', 'dataHash'{RESET}", file=sys.stderr)
        sys.exit(1)

    if not isinstance(salt_hex, str) or not re.match(r'^[0-9a-fA-F]+$', salt_hex) or len(salt_hex) % 2 != 0:
        print(f"{RED}[ERROR] Invalid salt: must be an even-length hexadecimal string{RESET}", file=sys.stderr)
        sys.exit(1)

    if not isinstance(expected_hash, str) or not re.match(r'^[0-9a-fA-F]{64}$', expected_hash):
        print(f"{RED}[ERROR] Invalid dataHash: must be a 64-character hexadecimal SHA3-256 string{RESET}", file=sys.stderr)
        sys.exit(1)

    print(f"{BOLD}1. Credential Subject & Attributes:{RESET}")
    print(f"  - Credential ID:    {CYAN}{credential_id}{RESET}")
    print(f"  - Subject:          {YELLOW}{raw_claim.get('subject', 'N/A')}{RESET}")
    print(f"  - Role / Degree:    {GREEN}{raw_claim.get('role', 'N/A')}{RESET}")
    print(f"  - Stored Salt:      {salt_hex}")
    print(f"  - Public Key ID:    {public_key_id}")
    print(f"  - Signature Algo:   {algorithm}")
    print(f"  - Ledger Anchor Tx: {anchor_tx}\n")

    # Step 1: RFC 8785 Canonicalization
    print(f"{BOLD}2. Step-by-Step Mathematical Verification:{RESET}")
    canonical_json = canonicalize(raw_claim)
    print("  [Step 1] RFC 8785 Canonical JSON:")
    print(f"           {CYAN}{canonical_json}{RESET}")

    # Step 2: Binary Salting Concatenation
    salt_bytes = bytes.fromhex(salt_hex)
    claim_bytes = canonical_json.encode('utf-8')
    payload = salt_bytes + claim_bytes
    print(f"  [Step 2] Salting Payload: Prepended {len(salt_bytes)}-byte CSPRNG salt")

    # Step 3: SHA3-256 Hash Commitment
    computed_hash = hashlib.sha3_256(payload).hexdigest()
    print("  [Step 3] Computed SHA3-256 Hash Commitment:")
    print(f"           {BOLD}{computed_hash}{RESET}")
    print("  [Step 4] Stored dataHash on Anchor Record:")
    print(f"           {BOLD}{expected_hash}{RESET}\n")

    # Step 4: Constant-time Hash Comparison (Level 1)
    hash_matches = hmac.compare_digest(computed_hash.lower(), expected_hash.lower())

    if not hash_matches:
        print(f"{BOLD}{RED}======================================================================{RESET}")
        print(f"{BOLD}{RED}  ✕ LEVEL 1 VERIFICATION FAILED: FORGERY OR TAMPERING DETECTED!{RESET}")
        print(f"{BOLD}{RED}======================================================================{RESET}")
        print(f"  {RED}Hash mismatch! The claim attributes or salt have been altered after issuance.{RESET}")
        print(f"  Expected Hash: {expected_hash}")
        print(f"  Computed Hash: {computed_hash}\n")
        sys.exit(1)

    print(f"  {GREEN}✓ Level 1 Passed:{RESET} Zero-Knowledge pre-image commitment is mathematically exact.\n")

    # Level 2: Signature & Public Key Verification
    print(f"{BOLD}3. Post-Quantum Signature Verification (ML-DSA-65):{RESET}")
    sig_verified = False
    sig_checked = False

    if signature_hex and public_key_hex:
        sig_checked = True
        try:
            if not isinstance(signature_hex, str) or not re.match(r'^[0-9a-fA-F]+$', signature_hex) or len(signature_hex) % 2 != 0:
                print(f"  {RED}✕ Level 2 Failed:{RESET} Signature must be a valid even-length hex string.")
                sig_verified = False
            elif not isinstance(public_key_hex, str) or not re.match(r'^[0-9a-fA-F]+$', public_key_hex) or len(public_key_hex) % 2 != 0:
                print(f"  {RED}✕ Level 2 Failed:{RESET} Public key must be a valid even-length hex string.")
                sig_verified = False
            else:
                sig_bytes = bytes.fromhex(signature_hex)
                pk_bytes = bytes.fromhex(public_key_hex)
                msg_bytes = bytes.fromhex(expected_hash)

                if len(sig_bytes) != 3309:
                    print(f"  {RED}✕ Level 2 Failed:{RESET} Signature length mismatch: expected 3,309 bytes (ML-DSA-65), got {len(sig_bytes)} bytes.")
                    sig_verified = False
                elif len(pk_bytes) != 1952:
                    print(f"  {RED}✕ Level 2 Failed:{RESET} Public key length mismatch: expected 1,952 bytes (ML-DSA-65), got {len(pk_bytes)} bytes.")
                    sig_verified = False
                else:
                    import oqs
                    with oqs.Signature("ML-DSA-65") as verifier:
                        sig_verified = verifier.verify(msg_bytes, sig_bytes, pk_bytes)

                    if sig_verified:
                        print(f"  {GREEN}✓ Level 2 Passed:{RESET} ML-DSA-65 post-quantum signature is authentic against provided Public Key.")
                    else:
                        print(f"  {RED}✕ Level 2 Failed:{RESET} Signature is INVALID for the provided Public Key.")
        except ImportError:
            # SECURITY: liboqs missing when signature data IS present is a hard failure.
            # Do NOT fall through to the success path — that would silently pass unverified credentials.
            print(f"\n{BOLD}{CYAN}======================================================================{RESET}")
            print(f"{BOLD}{YELLOW}  ⚠ VERIFICATION RESULT: DEPENDENCY MISSING — SIGNATURE NOT VERIFIED{RESET}")
            print(f"{BOLD}{CYAN}======================================================================{RESET}")
            print(f"  {RED}CRITICAL: `oqs` (liboqs-python) is not installed.{RESET}")
            print(f"  ML-DSA-65 post-quantum signature verification CANNOT proceed.")
            print(f"  The credential has NOT been cryptographically verified.")
            print(f"")
            print(f"  To perform full PQC signature verification, install liboqs-python:")
            print(f"    pip install liboqs-python")
            print(f"  Or run via the Docker container which bundles native liboqs:")
            print(f"    docker run --rm -v $PWD:/cred scatterid/verifier python3 tools/verify_offline.py /cred/credential.json")
            print(f"")
            print(f"  Exit code 2 = could not verify (dependency missing)")
            print(f"  Exit code 1 = signature invalid")
            print(f"  Exit code 0 = cryptographically verified\n")
            sys.exit(2)  # Distinct from exit(1)=invalid and exit(0)=verified
        except Exception as e:
            print(f"  {RED}✕ Level 2 Error:{RESET} Failed to verify signature: {e}")
            sig_verified = False
    else:
        print(f"  {YELLOW}[!] Notice:{RESET} 'signature' or 'publicKey' not supplied in credential bundle.")
        print(f"      (Pass `--public-key <hex>` to execute full PQC signature verification).")

    print(f"\n{BOLD}{CYAN}======================================================================{RESET}")
    if hash_matches and (not sig_checked or sig_verified):
        if sig_checked and sig_verified:
            print(f"{BOLD}{GREEN}  ✓ VERIFICATION RESULT: CRYPTOGRAPHICALLY VALID & AUTHENTIC{RESET}")
            print(f"{BOLD}{CYAN}======================================================================{RESET}")
            print(f"  {GREEN}The presented claim matches the exact zero-knowledge hash commitment.{RESET}")
            print(f"  {GREEN}ML-DSA-65 digital signature confirmed authentic from authorized issuer.{RESET}")
        else:
            print(f"{BOLD}{YELLOW}  ⚠ VERIFICATION RESULT: PRE-IMAGE COMMITMENT MATCH (UNAUTHENTICATED){RESET}")
            print(f"{BOLD}{CYAN}======================================================================{RESET}")
            print(f"  {YELLOW}The presented claim matches the hash commitment, but NO digital signature was verified.{RESET}")
            print(f"  {YELLOW}Anyone can construct arbitrary claims matching a self-generated commitment without issuer authority.{RESET}")
        print("  Zero-Knowledge Property Confirmed: Verification completed offline with zero leakage.")
        print(f"  {YELLOW}[!] Freshness Notice:{RESET} Offline verification confirms authenticity at issuance;")
        print(f"      it cannot confirm whether the credential has since been revoked on-chain.\n")
        sys.exit(0)
    else:
        print(f"{BOLD}{RED}  ✕ VERIFICATION RESULT: SIGNATURE VERIFICATION FAILED{RESET}")
        print(f"{BOLD}{CYAN}======================================================================{RESET}")
        sys.exit(1)

def main():
    print_header()
    parser = argparse.ArgumentParser(description="ScatterID Offline Cryptographic Verifier")
    parser.add_argument("file", nargs="?", help="Path to credential JSON file")
    parser.add_argument("--public-key", dest="public_key", help="Issuer ML-DSA-65 public key in hex format")

    args = parser.parse_args()

    if args.file and os.path.exists(args.file):
        with open(args.file, 'r', encoding='utf-8') as f:
            verify_offline(f.read(), cli_public_key=args.public_key)
    elif not sys.stdin.isatty():
        verify_offline(sys.stdin.read(), cli_public_key=args.public_key)
    else:
        parser.print_help()
        print("\nExample:")
        print("  python3 tools/verify_offline.py examples/credentials_input.json")
        sys.exit(0)

if __name__ == '__main__':
    main()

File: tools/verify_offline.js
#!/usr/bin/env node

/**
 * ============================================================================
 * ScatterID — "Don't Trust, Verify" Standalone Offline CLI Verifier (Node.js)
 * ============================================================================
 * This tool allows any third-party verifier, auditor, or relying party to
 * verify the cryptographic authenticity of an issued ScatterID credential
 * COMPLETELY OFFLINE using pure mathematics and standards:
 *   1. Zero-Knowledge Pre-image Commitment (RFC 8785 + FIPS 202 SHA3-256)
 *   2. NIST FIPS 204 ML-DSA-65 Signature Structure & Public Key Validation
 *
 * Ecosystem Role:
 *   - Node.js runtime for JavaScript / TypeScript developers (zero external dependencies).
 *   - For Python environments, use: python3 tools/verify_offline.py
 *
 * Usage:
 *   node tools/verify_offline.js <credential.json> [--public-key <hex>]
 *   cat credential.json | node tools/verify_offline.js
 * ============================================================================
 */

import fs from 'node:fs';
import { fileURLToPath } from 'node:url';
import { createHash, timingSafeEqual } from 'node:crypto';

/**
 * Validates whether a string contains an unescaped lone surrogate code point.
 */
function hasLoneSurrogate(value) {
  for (let i = 0; i < value.length; i++) {
    const code = value.charCodeAt(i);
    if (code >= 0xD800 && code <= 0xDBFF) {
      if (i === value.length - 1) return true;
      const next = value.charCodeAt(i + 1);
      if (!(next >= 0xDC00 && next <= 0xDFFF)) return true;
      i++;
    } else if (code >= 0xDC00 && code <= 0xDFFF) {
      return true;
    }
  }
  return false;
}

/**
 * Pure RFC 8785 JSON Canonicalization Scheme (JCS) implementation.
 * Guarantees zero external dependencies for offline verification:
 *   - Omits undefined and symbol properties in objects
 *   - Converts undefined and symbol values in arrays to null
 *   - Prohibits NaN and Infinity per RFC 8785 §3.2.2.3
 *   - Prohibits lone surrogates per RFC 8785 §3.2.2.2
 *   - Sorts object keys by UTF-16 code unit order per RFC 8785 §3.2.3
 *   - Normalizes negative zero (-0) to 0
 */
function canonicalize(object, seen = new Set()) {
  if (typeof object === 'number') {
    if (isNaN(object)) throw new Error('NaN is not allowed in RFC 8785 canonical JSON');
    if (!isFinite(object)) throw new Error('Infinity is not allowed in RFC 8785 canonical JSON');
    if (Object.is(object, -0)) return '0';
    return JSON.stringify(object);
  }
  if (typeof object === 'string') {
    if (hasLoneSurrogate(object)) throw new Error('Lone surrogate is not allowed in RFC 8785 canonical JSON');
    return JSON.stringify(object);
  }
  if (object === null || typeof object !== 'object') {
    return JSON.stringify(object);
  }
  if (typeof object.toJSON === 'function') {
    if (seen.has(object)) throw new Error('Circular reference detected in claim object');
    seen.add(object);
    const serialized = canonicalize(object.toJSON(), seen);
    seen.delete(object);
    return serialized;
  }
  if (seen.has(object)) throw new Error('Circular reference detected in claim object');
  seen.add(object);

  let result;
  if (Array.isArray(object)) {
    const values = object.map(item => {
      const val = (item === undefined || typeof item === 'symbol') ? null : item;
      return canonicalize(val, seen);
    });
    result = '[' + values.join(',') + ']';
  } else {
    const parts = [];
    for (const key of Object.keys(object).sort()) {
      if (object[key] === undefined || typeof object[key] === 'symbol') continue;
      parts.push(canonicalize(key) + ':' + canonicalize(object[key], seen));
    }
    result = '{' + parts.join(',') + '}';
  }
  seen.delete(object);
  return result;
}


// Terminal colors
const BOLD = '\x1b[1m';
const GREEN = '\x1b[32m';
const RED = '\x1b[31m';
const CYAN = '\x1b[36m';
const YELLOW = '\x1b[33m';
const RESET = '\x1b[0m';

function printHeader() {
  console.log(`${BOLD}${CYAN}======================================================================${RESET}`);
  console.log(`${BOLD}${CYAN}      ScatterID — Independent Offline Cryptographic Verifier (Node.js) ${RESET}`);
  console.log(`${BOLD}${CYAN}======================================================================${RESET}`);
  console.log(`Standards: RFC 8785 JSON Canonicalization (JCS) | SHA3-256 | ML-DSA-65 (FIPS 204)`);
  console.log(`Mode:      ${BOLD}100% OFFLINE (Zero Network Transit)${RESET}\n`);
}

function verifyOffline(credentialJsonStr, cliPublicKey) {
  let record;
  try {
    record = JSON.parse(credentialJsonStr);
  } catch (err) {
    console.error(`${RED}[ERROR] Input is not valid JSON: ${err.message}${RESET}`);
    process.exit(1);
  }

  // Support both wrapped output { credential: {...} } and direct credential objects
  const cred = record.credential || record;

  const rawClaim = cred.rawClaim || cred.claim;
  const saltHex = cred.salt;
  const expectedHash = cred.dataHash;
  const signatureHex = cred.signature;
  const publicKeyHex = cliPublicKey || cred.publicKey || cred.publicKeyHex;
  const publicKeyId = cred.publicKeyId || 'N/A';
  const credentialId = cred.credentialId || cred.id || 'N/A';
  const algorithm = cred.algorithm || 'ML-DSA-65 (NIST FIPS 204)';
  const anchorTxId = cred.anchorTxId || 'None';

  if (!rawClaim || !saltHex || !expectedHash) {
    console.error(`${RED}[ERROR] Incomplete credential object.${RESET}`);
    console.error(`Required fields: 'rawClaim' (or 'claim'), 'salt', 'dataHash'.`);
    process.exit(1);
  }

  if (typeof saltHex !== 'string' || !/^[0-9a-fA-F]+$/.test(saltHex) || saltHex.length % 2 !== 0) {
    console.error(`${RED}[ERROR] Invalid salt: must be an even-length hexadecimal string.${RESET}`);
    process.exit(1);
  }

  if (typeof expectedHash !== 'string' || !/^[0-9a-fA-F]{64}$/.test(expectedHash)) {
    console.error(`${RED}[ERROR] Invalid dataHash: must be a 64-character hexadecimal SHA3-256 string.${RESET}`);
    process.exit(1);
  }

  console.log(`${BOLD}1. Credential Subject & Attributes:${RESET}`);
  console.log(`  - Credential ID:    ${CYAN}${credentialId}${RESET}`);
  console.log(`  - Subject:          ${YELLOW}${rawClaim.subject || 'N/A'}${RESET}`);
  console.log(`  - Role / Degree:    ${GREEN}${rawClaim.role || 'N/A'}${RESET}`);
  console.log(`  - Stored Salt:      ${saltHex}`);
  console.log(`  - Public Key ID:    ${publicKeyId}`);
  console.log(`  - Signature Algo:   ${algorithm}`);
  console.log(`  - Ledger Anchor Tx: ${anchorTxId}\n`);

  // Step 1: RFC 8785 Canonicalization
  console.log(`${BOLD}2. Step-by-Step Mathematical Verification:${RESET}`);
  const canonicalJson = canonicalize(rawClaim);
  console.log(`  [Step 1] RFC 8785 Canonical JSON:`);
  console.log(`           ${CYAN}${canonicalJson}${RESET}`);

  // Step 2: Binary Payload Concatenation (Salt + Canonical JSON)
  const saltBytes = Buffer.from(saltHex, 'hex');
  const claimBytes = Buffer.from(canonicalJson, 'utf-8');
  const payload = Buffer.concat([saltBytes, claimBytes]);
  console.log(`  [Step 2] Salting Payload: Prepended ${saltBytes.length}-byte CSPRNG salt`);

  // Step 3: SHA3-256 Hash Computation
  const computedHash = createHash('sha3-256').update(payload).digest('hex');
  console.log(`  [Step 3] Computed SHA3-256 Hash Commitment:`);
  console.log(`           ${BOLD}${computedHash}${RESET}`);
  console.log(`  [Step 4] Stored dataHash on Anchor Record:`);
  console.log(`           ${BOLD}${expectedHash}${RESET}\n`);

  // Step 4: Constant-time Hash Comparison (Level 1)
  const compBuf = Buffer.from(computedHash, 'hex');
  const expBuf = Buffer.from(expectedHash, 'hex');

  const matches = (compBuf.length === expBuf.length) && timingSafeEqual(compBuf, expBuf);

  if (!matches) {
    console.log(`${BOLD}${RED}======================================================================${RESET}`);
    console.log(`${BOLD}${RED}  ✕ LEVEL 1 VERIFICATION FAILED: FORGERY OR TAMPERING DETECTED!${RESET}`);
    console.log(`${BOLD}${RED}======================================================================${RESET}`);
    console.log(`  ${RED}Hash mismatch! The claim attributes have been altered after signing.${RESET}`);
    console.log(`  Expected Hash: ${expectedHash}`);
    console.log(`  Computed Hash: ${computedHash}\n`);
    process.exit(1);
  }

  console.log(`  ${GREEN}✓ Level 1 Passed:${RESET} Zero-Knowledge pre-image commitment is mathematically exact.\n`);

  // Level 2: Signature & Public Key Structural Inspection
  console.log(`${BOLD}3. Post-Quantum Signature Verification (ML-DSA-65):${RESET}`);

  if (signatureHex && (typeof signatureHex !== 'string' || !/^[0-9a-fA-F]+$/.test(signatureHex) || signatureHex.length % 2 !== 0)) {
    console.error(`  ${RED}✕ Level 2 Failed: Signature must be an even-length hexadecimal string.${RESET}`);
    process.exit(1);
  }
  if (publicKeyHex && (typeof publicKeyHex !== 'string' || !/^[0-9a-fA-F]+$/.test(publicKeyHex) || publicKeyHex.length % 2 !== 0)) {
    console.error(`  ${RED}✕ Level 2 Failed: Public key must be an even-length hexadecimal string.${RESET}`);
    process.exit(1);
  }

  const sigBytes = signatureHex ? Buffer.from(signatureHex, 'hex') : null;
  const pkBytes = publicKeyHex ? Buffer.from(publicKeyHex, 'hex') : null;

  if (sigBytes) {
    console.log(`  - Signature Byte Length: ${sigBytes.length} bytes (ML-DSA-65 Standard: 3309 bytes)`);
    if (sigBytes.length === 3309) {
      // IMPORTANT: This is a structural pre-check only — NOT cryptographic verification.
      // A garbage or adversarially-crafted signature of the correct length will also pass this check.
      // Byte-length conformance is a fast-fail guard, never a substitute for verify().
      console.log(`  ${YELLOW}[i] Structural Pre-check:${RESET} Signature byte length matches ML-DSA-65 (3309 bytes). ${YELLOW}NOT a cryptographic verify() call.${RESET}`);
    } else {
      console.log(`  ${RED}✕ Structural Pre-check Failed:${RESET} Signature byte length (${sigBytes.length}B) differs from ML-DSA-65 standard (3309B).`);
    }
  } else {
    console.log(`  ${YELLOW}[!] Notice:${RESET} 'signature' not present in offline bundle.`);
  }

  if (pkBytes) {
    console.log(`  - Public Key Byte Length: ${pkBytes.length} bytes (ML-DSA-65 Standard: 1952 bytes)`);
    if (pkBytes.length === 1952) {
      // IMPORTANT: Same caveat — this is a byte-length pre-check, not a key validity proof.
      console.log(`  ${YELLOW}[i] Structural Pre-check:${RESET} Public key byte length matches ML-DSA-65 (1952 bytes). ${YELLOW}NOT a cryptographic verify() call.${RESET}`);
    } else {
      console.log(`  ${RED}✕ Structural Pre-check Failed:${RESET} Public key byte length (${pkBytes.length}B) differs from ML-DSA-65 standard (1952B).`);
    }
  }

  // Pure Node.js runtime has no native ML-DSA-65 (FIPS 204) engine without native liboqs bindings.
  // Structural validation passes, but mathematical signature verification must not be falsely asserted.
  console.log(`\n${BOLD}${CYAN}======================================================================${RESET}`);
  console.log(`${BOLD}${YELLOW}  ⚠ VERIFICATION RESULT: PRE-IMAGE COMMITMENT MATCH (UNAUTHENTICATED)${RESET}`);
  console.log(`${BOLD}${CYAN}======================================================================${RESET}`);
  console.log(`  ${GREEN}The presented claim matches the exact zero-knowledge hash commitment.${RESET}`);
  console.log(`  ${YELLOW}Notice: Pure Node.js CLI validates payload canonicalization and container structures.${RESET}`);
  console.log(`  ${YELLOW}        Cryptographic ML-DSA-65 signature verification requires liboqs bindings.${RESET}`);
  console.log(`  ${CYAN}To execute mathematical PQC signature verification, run:${RESET}`);
  console.log(`    python3 tools/verify_offline.py <credential.json>${publicKeyHex ? ` --public-key ${publicKeyHex}` : ''}`);
  console.log(`\n  Zero-Knowledge Property Confirmed: Verification completed offline with zero leakage.`);
  console.log(`  ${YELLOW}[!] Freshness Notice:${RESET} Offline verification confirms authenticity at issuance;`);
  console.log(`      it cannot confirm whether the credential has since been revoked on-chain.\n`);
  process.exit(0);
}

export { canonicalize, verifyOffline };

const isDirectRun = process.argv[1] && (
  fileURLToPath(import.meta.url) === fs.realpathSync(process.argv[1]) ||
  process.argv[1].endsWith('verify_offline.js')
);

if (isDirectRun) {
  printHeader();

  const args = process.argv.slice(2);
  let filePath = null;
  let pubKey = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--public-key' && args[i + 1]) {
      pubKey = args[i + 1];
      i++;
    } else if (!filePath) {
      filePath = args[i];
    }
  }

  if (filePath && fs.existsSync(filePath)) {
    const content = fs.readFileSync(filePath, 'utf8');
    verifyOffline(content, pubKey);
  } else if (!process.stdin.isTTY) {
    let content = '';
    process.stdin.on('data', chunk => { content += chunk; });
    process.stdin.on('end', () => verifyOffline(content, pubKey));
  } else {
    console.log(`Usage:`);
    console.log(`  node tools/verify_offline.js <path-to-credential.json> [--public-key <hex>]`);
    console.log(`  cat credential.json | node tools/verify_offline.js\n`);
    console.log(`Example:`);
    console.log(`  node tools/verify_offline.js examples/credentials_input.json\n`);
    process.exit(0);
  }
}
